html/admin/api/server/assetImport.php

```diff
@@ -0,0 +1,175 @@
+    $assetType = $DBLIB->getOne("assetTypes", ["assetTypes.assetTypes_id"]);
+
+    if(!$assetType){
+        //Create new asset type
+        //Manufacturer
+        if($row[8] == "") $row[8] = "Unknown/Generic"; //Use the generic manufacturer if none specified
+        $DBLIB->where("manufacturers_name", "%" . $row[8] . "%", "LIKE");
+        $DBLIB->where("(manufacturers.instances_id = ? or manufacturers.instances_id IS NULL)", [$_POST['instances_id']]);
+        $manufacturer = $DBLIB->getOne("manufacturers", ["manufacturers.manufacturers_id"]);
+        if (!$manufacturer) {
+            $manufacturer = [
+                "manufacturers_name" => $row[8],
+                "instances_id" => $_POST['instances_id']
+            ];
+            $manufacturer['manufacturers_id'] = $DBLIB->insert("manufacturers", $manufacturer);
+        }
+        
+        //Asset Category
+        $DBLIB->where("assetCategories_id", $row[7]);
+        $DBLIB->where("(assetCategories.instances_id = ? or assetCategories.instances_id IS NULL)", [$_POST['instances_id']]);
+        $DBLIB->where("assetCategories.assetCategories_deleted", 0);
+        $assetCategory = $DBLIB->getOne("assetCategories", ["assetCategories.assetCategories_id"]);
+        if (!$assetCategory) {
+            //Asset Category not found
+            //This is the one thing we can't just create with data from the CSV
+            array_push($failedAssets, ["tag" => $row[9], "reason" => "Asset Category with ID ". $row[7] . " not found"]);
+            continue;
+        }
+
+        //Map definable fields
+        $definableFields = "";
+        for ($j = 17; $j < 27; $j++) {
+            $definableFields .= $row[$j] . ",";
+        }
+        $definableFields = rtrim($definableFields, ",");
+
+         
+        //Actually create new asset type
+        $assetType = [
+            "assetTypes_name" => $row[0],
+            "assetCategories_id" => $row[7],
+            "manufacturers_id" => $manufacturer['manufacturers_id'],
+            "instances_id" => $_POST['instances_id'],
+            "assetTypes_description" => $row[1],
+            "assetTypes_productLink" => $row[2],
+            "assetTypes_definableFields" => $definableFields,
+            "assetTypes_mass" => $row[3],
+            "assetTypes_inserted" => time(),
+            "assetTypes_dayRate" => $row[4],
+            "assetTypes_weekRate" => $row[5],
+            "assetTypes_value" => $row[6],   
+        ];
+        $assetType['assetTypes_id'] = $DBLIB->insert("assetTypes", $assetType);
+        if ($assetType['assetTypes_id']) array_push($createdAssetTypes, $assetType);
+        else array_push($failedAssets, ["tag" => $row[9], "reason" => "Error creating Asset Type"]);
+
+    }
+
+    //Asset Tag
+    if (!isset($row[9]) or $row[9] == null) $row[9] = generateNewTag();
+    else {
+        //check Tag is not going to be duplicated
+        $DBLIB->where("assets.instances_id",$AUTH->data['instance']['instances_id']);
```

Comment:
Is this not the same as line 75?

Reply:
This checks for the asset not asset type, so you can't import a duplicate asset tag